### ISSUE TITLE
Feature request: a way to set/get default ticketer dynamically

### ISSUE THREAD

**Issue by Alvenix** [rustls/rustls#1876]
**Checklist**
* [x] I've searched the issue tracker for similar requests

**Is your feature request related to a problem? Please describe.**
Currently I am trying to make Rocket use the default CryptoProvider if it is set. [This](https://github.com/rwf2/Rocket/pull/2758#issuecomment-2021688150) is the only blocking issue. 

**Describe the solution you'd like**
Either ticketer is added as a field on CryptoProvider or add a way to set/get default ticketer separately.

**Describe alternatives you've considered**
I considered simply using ring Ticketer but this may be unexpected to the users who change CryptoProvider.

**Comment by cpu** [comment]
I'm not sure that we want to link the `ProducesTicketer` implementation into the `CryptoProvider` struct as a field but I'm curious what others think. :thinking:  

There's not really a "default" ticketer from my perspective, just two convenient implementations of `dyn ProducesTickets` made available for the built in cryptography provider options. From that viewpoint I think what's missing is a way to get the right one depending on whether you have unambiguously set crate features such that it's possible to pick the right one without more thought.

So, what about something inspired by `CryptoProvider::from_crate_features` like:

```rust
impl ServerConfig {
  ...
  pub fn ticketer_from_crate_features() -> Option<Result<Arc<dyn ProducesTickets>, Error>> {
        #[cfg(all(feature = "ring", not(feature = "aws_lc_rs")))]
        {
            return Some(rustls::crypto::ring::Ticketer::new());
        }

        #[cfg(all(feature = "aws_lc_rs", not(feature = "ring")))]
        {
            return Some(aws_lc_rs::default_provider());
        }
        
        #[cfg(not(feature = "aws_lc_rs"),  not(feature = "ring"))]
        {
            return None
        }
  }
}
```

For other cases, where the features aren't unambigious, or an external-to-rustls `CryptoProvider` has been used, I think the expectation is that the user setting the features/provider constructs a `dyn ProducesTickets` impl custom to their situation and sets it as the `ServerConfig.ticketer`. Neither the aws-lc-rs or ring ticketer is appropriate in that case.

**Comment by cpu** [comment]
>  I had too many panics due to enabling the two features together.

Right, if the two options are enabled together, and no explicit default has been set, and someone uses a builder fn that uses the default it will panic rather than try to guess which of the two to use. I don't think we should try to do something different here by offering a parameter.

If the user has a working configuration with both features enabled irrespective of ticketer config it's because something is constructing the ring or aws provider, installing it as default, and then using the builders. In this case whatever is doing so has the necessary knowledge to also construct a choice of the aws or ring ticketer and install it to the server config field, no?

IMO it's only when the default is implied by the use of **unambiguous** crate features that there's not enough knowledge to know which of the ticketers to construct and that's what the helper I proposed would offer.

**Comment by cpu** [comment]
> Similarly, if the same could be applied to CryptoProvider::from_crate_features, as I feel its current API can be too footguny.

I think changes to how default crypto providers are constructed is a separate topic that should be its own issue/discussion (_but I would also caution that the current design is the result of a fair amount of iteration and I don't think we're especially keen to revisit core aspects without strong motivation for why something else is a better choice_).

**Comment by djc** [comment]
(Going to ignore the #1877 discussion in my comments here, thanks @Alvenix for splitting that off.)

So I guess it would be cool if we could reformulate the `ring_like` `AeadTicketer` (which somewhat confusingly lives in `crypto::ring::ticketer`) to rely on, maybe, a `Tls13AeadAlgorithm` implementation instead of using the `ring_like::aead` primitives directly?

In the absence of that, maybe we should come to consensus in #1877 before we resolve the details of the interface needed to expose a ticketer?

**Comment by zh-jq-b** [comment]
I also want to get a default Ticketer from the default CryptoProvider, is it possible to do this?

May be the `KeyProvider` trait can be extended to include
```rust
fn new_provided_ticketer() -> Result<Arc<dyn ProducesTickets>, Error> {
    Ok(Arc::new(handy::NeverProducesTickets {}))
}
```

**Comment by ctz** [comment]
> I also want to get a default Ticketer from the default CryptoProvider, is it possible to do this?
> 
> May be the `KeyProvider` trait can be extended to include
> 
> ```rust
> fn new_provided_ticketer() -> Result<Arc<dyn ProducesTickets>, Error> {
>     Ok(Arc::new(handy::NeverProducesTickets {}))
> }
> ```

I think this seems like a good option.

**Comment by djc** [comment]
Implemented in

- #2754
